### ISSUE TITLE
Bot fails an connection and dies

### ISSUE THREAD

**Issue by Mr-Rixa** [mastercake10/TelegramChat#6]
Not entirely sure what happened here, but after this you'll need to restart server to get the bot to connect back on telegram.

Censored bot apikey
Running version v1.0.9.9 of TelegramChat
server is running CraftBukkit version git-Spigot-549c1fa-45c8386

```
[03:30:34 WARN]: java.io.IOException: Server returned HTTP response code: 502 for URL: https://api.telegram.org/botxxxxxxxxx:yyyyyyyyyyyy/getUpdates?offset=68771061
[03:30:34 WARN]:        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1876)
[03:30:34 WARN]:        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1474)
[03:30:34 WARN]:        at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:254)
[03:30:34 WARN]:        at de.Linus122.TelegramChat.Telegram.sendGet(Telegram.java:193)
[03:30:34 WARN]:        at de.Linus122.TelegramChat.Telegram.getUpdate(Telegram.java:56)
[03:30:34 WARN]:        at de.Linus122.TelegramChat.Main$1.run(Main.java:73)
[03:30:34 WARN]:        at org.bukkit.craftbukkit.v1_12_R1.scheduler.CraftTask.run(CraftTask.java:71)
[03:30:34 WARN]:        at org.bukkit.craftbukkit.v1_12_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:353)
[03:30:34 WARN]:        at net.minecraft.server.v1_12_R1.MinecraftServer.D(MinecraftServer.java:739)
[03:30:34 WARN]:        at net.minecraft.server.v1_12_R1.DedicatedServer.D(DedicatedServer.java:406)
[03:30:34 WARN]:        at net.minecraft.server.v1_12_R1.MinecraftServer.C(MinecraftServer.java:679)
[03:30:34 WARN]:        at net.minecraft.server.v1_12_R1.MinecraftServer.run(MinecraftServer.java:577)
[03:30:34 WARN]:        at java.lang.Thread.run(Thread.java:748)
[03:30:34 INFO]: [Telegram] Sorry, but could not connect to Telegram servers. The token could be wrong.
[03:30:34 WARN]: [TelegramChat] Task #8 for TelegramChat v1.0.9.9 generated an exception
java.lang.NullPointerException: null
        at de.Linus122.TelegramChat.Telegram.getUpdate(Telegram.java:64) ~[?:?]
        at de.Linus122.TelegramChat.Main$1.run(Main.java:73) ~[?:?]
        at org.bukkit.craftbukkit.v1_12_R1.scheduler.CraftTask.run(CraftTask.java:71) ~[spigot.jar:git-Spigot-549c1fa-45c8386]
        at org.bukkit.craftbukkit.v1_12_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:353) [spigot.jar:git-Spigot-549c1fa-45c8386]
        at net.minecraft.server.v1_12_R1.MinecraftServer.D(MinecraftServer.java:739) [spigot.jar:git-Spigot-549c1fa-45c8386]
        at net.minecraft.server.v1_12_R1.DedicatedServer.D(DedicatedServer.java:406) [spigot.jar:git-Spigot-549c1fa-45c8386]
        at net.minecraft.server.v1_12_R1.MinecraftServer.C(MinecraftServer.java:679) [spigot.jar:git-Spigot-549c1fa-45c8386]
        at net.minecraft.server.v1_12_R1.MinecraftServer.run(MinecraftServer.java:577) [spigot.jar:git-Spigot-549c1fa-45c8386]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_141]
```

**Comment by mastercake10** [comment]
A 502 error is very uncommon, so I didn't treat it. But I agree that the bot should try to reconnect instead of crashing the plugin.

**Comment by Programie** [comment]
I just got the same issue.

It is really bad to see that a plugin like this one just stopped working in the middle of the night and you have to manually reload it to get it working again.

An HTTP status code 502 (Bad Gateway) is usually returned by a reverse proxy not able to reach the servers behind it. This is typically a temporary problem (network issue between proxy and server, server behind proxy just got restarted, etc.) and the client should retry after a while.

All 5xx HTTP status codes signal a server problem and not a client problem. If the Telegram API is correctly using HTTP status codes it should return a 4xx HTTP status code in case of a client problem (e.g. invalid token). So for most 5xx HTTP status codes, you could simply try to reconnect to the Telegram API (with a short delay to prevent killing the Telegram API - maybe the 5xx response is caused by overload).
I would implement the delay to increase after each failed retry and then give up after retrying for too long. So for example the plugin could retry after 5 seconds, 15 seconds, 30 seconds, 1 minute, 5 minutes, 15 minutes, 30 minutes, 1 hour, 3 hours, 6 hours and then finally give up.

**Comment by mastercake10** [comment]
fixed in 2cbd9242b3d535f0aa28f54c7b3772cc82f823ac
